metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+
+    def get_report(self, workspace: Workspace, report_id: str) -> Optional[Report]:
+        reports: List[Report] = self.get_reports(
+            workspace, _filter=f"id eq '{report_id}'"
+        )
+
+        if len(reports) == 0:
+            return None
+
+        return reports[0]
+
+    def get_tiles(self, workspace: Workspace, dashboard: Dashboard) -> List[Tile]:
+
+        """
+        Get the list of tiles from PowerBi for the given workspace identifier
+
+        TODO: Pagination. As per REST API doc (https://docs.microsoft.com/en-us/rest/api/power-bi/dashboards/get
+        -tiles), there is no information available on pagination
+
+        """
+
+        def new_dataset_or_report(tile_instance: Any) -> dict:
+            """
+            Find out which is the data source for tile. It is either REPORT or DATASET
+            """
+            report_fields = {
+                "dataset": (
+                    workspace.datasets.get(tile_instance.get("datasetId"))
+                    if tile_instance.get("datasetId") is not None
+                    else None
+                ),
+                "report": (
+                    self.get_report(
+                        workspace=workspace,
+                        report_id=tile_instance.get("reportId"),
+                    )
+                    if tile_instance.get("reportId") is not None
+                    else None
+                ),
+                "createdFrom": Tile.CreatedFrom.UNKNOWN,
+            }
+
+            # Tile is either created from report or dataset or from custom visualization
+            if report_fields["report"] is not None:
+                report_fields["createdFrom"] = Tile.CreatedFrom.REPORT
+            elif report_fields["dataset"] is not None or (
+                report_fields["dataset"] is None
+                and tile_instance.get("datasetId") is not None
+            ):  # Admin API is disabled and hence dataset instance is missing
+                report_fields["createdFrom"] = Tile.CreatedFrom.DATASET
+            else:
+                report_fields["createdFrom"] = Tile.CreatedFrom.VISUALIZATION
+
+            title: Optional[str] = tile_instance.get("title")
+            _id: Optional[str] = tile_instance.get("id")
+            created_from: Any = report_fields["createdFrom"]
+            logger.info(f"Tile {title}({_id}) is created from {created_from}")
+
+            return report_fields
+
+        tile_list_endpoint: str = self.get_tiles_endpoint(
+            workspace, dashboard_id=dashboard.id
+        )
+        # Hit PowerBi
+        logger.info(f"Request to URL={tile_list_endpoint}")
+        response = self._request_session.get(
+            tile_list_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+        response.raise_for_status()
+
+        # Iterate through response and create a list of PowerBiAPI.Dashboard
+        tile_dict: List[Any] = response.json()[Constant.VALUE]
```

Comment:
Question - what happens if these JSON response don't contain a "value" field? Seems we will just blow up with a runtime exception

Reply:
replaced with `response.json().get(Constant.VALUE, [])`